scene/resources/animation.cpp

```diff
@@ -4048,83 +4048,123 @@ void Animation::clear() {
 	emit_signal(SceneStringNames::get_singleton()->tracks_changed);
 }
 
-bool Animation::_vector3_track_optimize_key(const TKey<Vector3> t0, const TKey<Vector3> t1, const TKey<Vector3> t2, real_t p_allowed_velocity_err, real_t p_allowed_angular_error, real_t p_allowed_precision_error) {
+bool Animation::_float_track_optimize_key(const TKey<float> t0, const TKey<float> t1, const TKey<float> t2, real_t p_allowed_velocity_err, real_t p_allowed_precision_error) {
 	// Remove overlapping keys.
 	if (Math::is_equal_approx(t0.time, t1.time) || Math::is_equal_approx(t1.time, t2.time)) {
 		return true;
 	}
-	if ((t0.value - t1.value).length() < p_allowed_precision_error && (t1.value - t2.value).length() < p_allowed_precision_error) {
+	if (abs(t0.value - t1.value) < p_allowed_precision_error && abs(t1.value - t2.value) < p_allowed_precision_error) {
 		return true;
 	}
 	// Calc velocities.
-	Vector3 vc0 = (t1.value - t0.value) / (t1.time - t0.time);
-	Vector3 vc1 = (t2.value - t1.value) / (t2.time - t1.time);
-	real_t v0 = vc0.length();
-	real_t v1 = vc1.length();
+	double v0 = (t1.value - t0.value) / (t1.time - t0.time);
+	double v1 = (t2.value - t1.value) / (t2.time - t1.time);
 	// Avoid zero div but check equality.
 	if (abs(v0 - v1) < p_allowed_precision_error) {
 		return true;
 	} else if (abs(v0) < p_allowed_precision_error || abs(v1) < p_allowed_precision_error) {
 		return false;
 	}
-	// Check axis.
-	if (vc0.normalized().dot(vc1.normalized()) >= 1.0 - p_allowed_angular_error * 2.0) {
-		real_t ratio = v0 < v1 ? v0 / v1 : v1 / v0;
+	if (!signbit(v0 * v1)) {
+		v0 = abs(v0);
+		v1 = abs(v1);
+		double ratio = v0 < v1 ? v0 / v1 : v1 / v0;
```

Comment:
In the old optimizer, it does not get the absolute value here, and the positive and negative values are reversed, causing the ratio to be incorrect.